Lib/functools.py

```diff
@@ -867,7 +878,8 @@ def register(cls, func=None):
             # only import typing if annotation parsing is necessary
             from typing import get_type_hints
             argname, cls = next(iter(get_type_hints(func).items()))
-            if not isinstance(cls, type) and not _is_valid_union_type(cls):
+            if not ((isinstance(cls, type) and not isinstance(cls, GenericAlias))
```

Comment:
This is repeated logic from lines 858-859. Could this be factored out into an `_is_valid_dispatch_type` function? I think that would make it easier to modify if we decide to support other types in `singledispatch` in the future.

Reply:
Good idea. Done.